pkg/config/main.go

```diff
@@ -0,0 +1,338 @@
+		ps = toPolicies(a.ExportPolicyList)
+		apiServer.SetPolicyAssignment(context.Background(), &api.SetPolicyAssignmentRequest{
+			Assignment: server.NewAPIPolicyAssignmentFromTableStruct(&table.PolicyAssignment{
+				Name:     table.GLOBAL_RIB_NAME,
+				Type:     table.POLICY_DIRECTION_EXPORT,
+				Policies: ps,
+				Default:  def,
+			}),
+		})
+
+		updatePolicy = true
+	}
+	c = newConfig
+	addPeerGroup(apiServer, addedPg)
+	deletePeerGroup(apiServer, deletedPg)
+	updatePolicy = updatePolicy || updatePeerGroup(apiServer, updatedPg)
+	addDynamicNeigbors(apiServer, newConfig.DynamicNeighbors)
+	addNeighbors(apiServer, added)
+	deleteNeighbors(apiServer, deleted)
+	updatePolicy = updatePolicy || updateNeighbors(apiServer, updated)
+	if updatePolicy {
+		if _, err := apiServer.ResetPeer(context.Background(), &api.ResetPeerRequest{
+			Address:   "",
+			Direction: api.ResetPeerRequest_IN,
+			Soft:      true,
+		}); err != nil {
+			log.Warn(err)
+		}
+	}
+	return c
+}
+
+func addNeighbors(apiServer *server.Server, added []config.Neighbor) {
+	for _, p := range added {
+		log.Infof("Peer %v is added", p.State.NeighborAddress)
+		if _, err := apiServer.AddPeer(context.Background(), &api.AddPeerRequest{
+			Peer: server.NewPeerFromConfigStruct(&p),
+		}); err != nil {
+			log.Warn(err)
+		}
+	}
+}
+
+func deleteNeighbors(apiServer *server.Server, deleted []config.Neighbor) {
+	for _, p := range deleted {
+		log.Infof("Peer %v is deleted", p.State.NeighborAddress)
+		if _, err := apiServer.DeletePeer(context.Background(), &api.DeletePeerRequest{
+			Address: p.State.NeighborAddress,
+		}); err != nil {
+			log.Warn(err)
+		}
+	}
+}
+
+func updateNeighbors(apiServer *server.Server, updated []config.Neighbor) bool {
+	updatePolicy := false
+	for _, p := range updated {
+		log.Infof("Peer %v is updated", p.State.NeighborAddress)
+		if u, err := apiServer.UpdatePeer(context.Background(), &api.UpdatePeerRequest{
+			Peer: server.NewPeerFromConfigStruct(&p),
+		}); err != nil {
+			log.Warn(err)
+		} else {
+			updatePolicy = updatePolicy || u.NeedsSoftResetIn
+		}
+	}
+	return updatePolicy
+}
+
+func addPeerGroup(apiServer *server.Server, added []config.PeerGroup) {
+	for _, pg := range added {
+		log.Infof("PeerGroup %s is added", pg.Config.PeerGroupName)
+		if _, err := apiServer.AddPeerGroup(context.Background(), &api.AddPeerGroupRequest{
+			PeerGroup: server.NewPeerGroupFromConfigStruct(&pg),
+		}); err != nil {
+			log.Warn(err)
+		}
+	}
+}
+
+func deletePeerGroup(apiServer *server.Server, deleted []config.PeerGroup) {
+	for _, pg := range deleted {
+		log.Infof("PeerGroup %s is deleted", pg.Config.PeerGroupName)
+		if _, err := apiServer.DeletePeerGroup(context.Background(), &api.DeletePeerGroupRequest{
+			Name: pg.Config.PeerGroupName,
+		}); err != nil {
+			log.Warn(err)
+		}
+	}
+}
+
+func updatePeerGroup(apiServer *server.Server, updated []config.PeerGroup) bool {
+	updatePolicy := false
+	for _, pg := range updated {
```

Comment:
This method appears to do some redundant work. This was not a mistake in refactoring. The original does the same redundant work. I only did pure refactoring in this PR and did not attempt to fix anything that I thought was incorrect. Decomposing this code into smaller functions helps to see this kind of thing.